Fix Owntracks Accuracy Issue

HydrelioxGitHub opened the pull request:
**Description:**
Dismiss a GPS report if this one is not accurate enough

**Related issue (if applicable):** #1270

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
device_tracker:
  platform: owntracks
  max_gps_accuracy: 200
```

**Checklist:**
- [x] Local tests with `tox` run successfully.
- [x] TravisCI does not fail. **Your PR cannot be merged unless CI is green!**
- [x] [Fork is up to date](http://stackoverflow.com/a/7244456) and was rebased on the `dev` branch before creating the PR.
- [x] Commits have been [squashed](https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit).
- If code communicates with devices:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
  - [ ] New dependencies are only imported inside functions that use them ([example](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.
- If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

sander76 committed:
first commit

sander76 committed:
flake8 correction

sander76 committed:
- removed update method
- removed failing tests from test_scene

sander76 committed:
removed unnecessary properties.

sander76 committed:
1. added platform per scene entry.
2. changed homeassistant scene setup_platform method to work with 1.

sander76 committed:
removed logger.
I guess I should add some error checking, but I'd like you to okay this code first.

sander76 committed:
removed logger.
I guess I should add some error checking, but I'd like you to okay this code first.

fabaff committed:
Add MQTT lock support

sander76 committed:
Added your suggestions.
Looking at your code suggestion below I am not sure exactly how other people might want to put in lists. (But I am missing a more general overview of the code)

 ```
if not isinstance(scene_config,list):
  scene_config=[scene_config]
```

But it is there !
And changed "config" to "states" !

Commit:
Added tests for rfxtrx sensor

HydrelioxGitHub committed:
Fix Accuracy

pavoni commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -78,6 +91,16 @@ def owntracks_event_update(topic, payload, qos):
         if not isinstance(data, dict) or data.get('_type') != 'transition':
             return
 
+        # Check for GPS accuracy
```
Owntracks will only send a single 'enter' and 'leave' event - so you can't ignore them.

It doesn't make sense to check accuracy during an 'enter' event in any case, because Owntracks already made the decision - and HA just does as it's told.

We could check accuracy when switching back to gps after processing a leave event, but it's difficult to do correctly since we need to exit the region.

I think I'd suggest not worrying about accuracy when using owntracks regions.

pavoni commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -48,6 +51,16 @@ def owntracks_location_update(topic, payload, qos):
             return
 
+        # Check for GPS accuracy
+        if CONF_MAX_GPS_ACCURACY in config:
```
Probably best to set this  up in the scope of setup_scanner rather than in the callback function called on every update.

HydrelioxGitHub commented:
You suggest to just filter on a location update and not on a event update ?

pavoni commented:
That's what seems safest to me.

To implant accuracy for  events you need to change owntracks apps!

The only accuracy you could safely check in events is when it exits to gps, but you'd still need to 'see' the exit region.

HydrelioxGitHub commented:
I will correct the code as you suggest, However, I hope it will fix the #1270  issue

pavoni commented:
Events are only sent if you set them up in own tracks.

You can ignore inaccurate 'exit to gps' events if you want to, you'd need to duplicate and move some code to make sure the 'see' is always called when exiting to a region and only skipped when 'exiting to gps'.

pavoni commented:
Could you also add a couple of extra tests to check the new boundary conditions?

kk7ds committed:
Give mFi options to control TLS usage

The default configuration of the mFi controller generates self-signed
certificates which are valid for short periods of time, and which are
regnerated on start or as needed. This makes requests mad. Since most
people will be using the self-signed certificates anyway, add options
to let them choose non-TLS, or unverified connections if they want/need.

This bumps the mficlient requirement to 0.3.0 for proper handling of
verify=False.

HydrelioxGitHub committed:
Merge branch 'dev' of https://github.com/balloob/home-assistant into Owntracks-GPS-Accuracy

HydrelioxGitHub committed:
Do not take accuracy in account when event are fired

balloob commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -28,10 +28,18 @@
 LOCK = threading.Lock()
 
+CONF_MAX_GPS_ACCURACY = 'max_gps_accuracy'
+DEFAULT_MAX_GPS_ACCURACY = 500
```
Let's make the default set to `None` to not lose backwards compatibility.

balloob commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -28,10 +28,18 @@
     """ Set up an OwnTracks tracker. """
 
+    if CONF_MAX_GPS_ACCURACY in config:
+        max_gps_accuracy = config[CONF_MAX_GPS_ACCURACY]
```
No `if` statement needed if you use `get` on a `dict`. It will return `None` (or the optional second argument) if not found.
```python
max_gps_accuracy = config.get(CONF_MAX_GPS_ACCURACY)
```

balloob committed:
Update Dockerfile: link Z-Wave config

HydrelioxGitHub committed:
Changes to not lose backwards compatibility

HydrelioxGitHub commented:
I will wait several hours before writing tests because I want clean code. Maybe if there is another comment, code will change and tests written before will be useless

kk7ds committed:
Track WeMo devices by serial number

This makes us track WeMo devices by serial number instead of URL. Per
@jaharkes' suggestion, tracking via URL could potentially be a problem
if a device changes to another IP and then appears to be discovered again.
Since we can't track based on mac (for static-configured devices), we
can use the serial number exposed to make sure we're always looking at
a list of unique devices.

balloob commented on homeassistant/components/device_tracker/owntracks.py:
```diff
@@ -48,6 +52,12 @@ def owntracks_location_update(topic, payload, qos):
             return
 
+        # Check for GPS accuracy
+        if ('acc' in data) and (max_gps_accuracy is not None):
```
Can you merge these two `if`statements into 1 ?

pavoni commented:
Have been thinking about checking accuracy on region exit. If you wanted to do this you could do something like.

````
                if new_region:
...
                    _LOGGER.info("Exit to %s", new_region)
                    see(**kwargs)
                    see_beacons(dev_id, kwargs)
                else:
                    _LOGGER.info("Exit to GPS")
                    # Check for GPS accuracy
                    if not ('acc' in data and 
                            max_gps_accuracy is not None and
                            data['acc'] > max_gps_accuracy):

                        see(**kwargs)
                        see_beacons(dev_id, kwargs)
                else:
                    _LOGGER.info("Inaccurate GPS reported")

                beacons = MOBILE_BEACONS_ACTIVE[dev_id]
````

pavoni commented:
Looks good otherwise. Perhaps just add a couple of tests (or a couple more if you decide to add the leave region code).

HydrelioxGitHub commented:
A question as a Python newbie. You can use `'acc' in data`

pavoni commented:
That's right, detail here https://docs.python.org/2/library/stdtypes.html#boolean-operations-and-or-not

pavoni commented:
Or you could do something like `dict.get('a', 1)` which will return the value of 'dict['a']` or 1 if it was missing.

HydrelioxGitHub committed:
Merge statement to filter GPS report

HydrelioxGitHub committed:
Filter inaccurate GPS report in event update

HydrelioxGitHub commented:
@pavoni I don't know if I correctly integrate your code...

jaharkes committed:
Update netdisco to 0.5.4

pavoni commented:
@HydrelioxGitHub Looks fine to me. Let me know if you need help with the tests.

Commit:
Added tests for RFxtrx device, updated rfxtrx to ver 0.5, fixed bug in setting brightness for rfxtrx light

kk7ds committed:
Fix SNMP device_tracker results without MAC addresses

@llauren reported that his router returns empty MAC addresses for things
on different subnets, which causes heartache for the HA snmp device tracker.

HydrelioxGitHub committed:
Add Test for an inaccurate location report

HydrelioxGitHub committed:
I screwed up myself

HydrelioxGitHub commented:
I think I need help :sos:

pavoni commented on tests/components/device_tracker/test_owntracks.py:
```diff
@@ -169,6 +190,15 @@ def test_location_update(self):
         self.assert_location_state('outer')
 
+    def test_location_inaccurate_gps(self):
+        self.hass.config.CONF_MAX_GPS_ACCURACY = 200
```
I think your issue is that owntracks reads this value during `setup_scanner` - which has been run in the `setup_method` which is run before the test. 

So you could either change the core set-up with something like:-
````
    def setup_method(self, method):
        """ Init needed objects. """
        self.hass = get_test_home_assistant()
        mock_mqtt_component(self.hass)
        self.assertTrue(device_tracker.setup(self.hass, {
            device_tracker.DOMAIN: {
                CONF_PLATFORM: 'owntracks',
                CONF_MAX_GPS_ACCURACY: 200
            }}))
````
This will change the config for all tests, which I think is fine.

pavoni commented:
Would also suggesting adding a event exit test, something like this (untested)
````
    def test_event_entry_exit_inaccurate(self):
        self.send_message(EVENT_TOPIC, REGION_ENTER_MESSAGE)

        # Enter uses the zone's gps co-ords
        self.assert_location_latitude(2.1)
        self.assert_location_accuracy(10.0)
        self.assert_location_state('inner')

        self.send_message(EVENT_TOPIC, REGION_LEAVE_INACCURATE_MESSAGE)

        # Exit doesn't use inaccurate gps 
        self.assert_location_latitude(2.1)
        self.assert_location_accuracy(10.0)
        self.assert_location_state('inner')

        # But does exit region correctly
        self.assertFalse(owntracks.REGIONS_ENTERED[USER])
````

Commit:
update rfxtrx lib to ver 0.5

HydrelioxGitHub committed:
Fix update location test

HydrelioxGitHub committed:
Add Test for exiting region with inaccurate gps  report

HydrelioxGitHub committed:
Fixing Tests

Commit:
Update pychromecast 0.7.2

pavoni commented on homeassistant/components/frontend/www_static/home-assistant-polymer:
```diff
@@ -1 +0,0 @@
-Subproject commit 4824268203068dfbd03e9d5ff5f4b59b04f3fe67
```
Could you remove this?

pavoni commented:
If you can merge the commits - and remove the subproject commit, this looks good to go :+1: 

Thanks!

HydrelioxGitHub commented:
I've created a new branch to make a clean PR.
